src/Security/Guard/JwtGuardAuthenticator.php

```diff
@@ -0,0 +1,174 @@
+use Auth0\SDK\Exception\CoreException;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\User\User;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
+
+/**
+ * Handles authentication with JSON Web Tokens through the 'Authorization' request header.
+ *
+ * @author Niels Nijens <nijens.niels@gmail.com>
+ */
+class JwtGuardAuthenticator extends AbstractGuardAuthenticator
+{
+    /**
+     * @var Auth0Service
+     */
+    private $authZeroService;
+
+    /**
+     * Constructs a new JwtGuardAuthenticator instance.
+     *
+     * @param Auth0Service $authZeroService
+     */
+    public function __construct(Auth0Service $authZeroService)
+    {
+        $this->authZeroService = $authZeroService;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request)
+    {
+        return $request->headers->has('Authorization');
+    }
+
+    /**
+     * Retrieves the authentication credentials from the 'Authorization' request header.
+     *
+     * @param Request $request
+     *
+     * @return array|null
+     */
+    public function getCredentials(Request $request)
+    {
+        // Removes the 'Bearer ' part from the Authorization header value.
+        $jwt = substr($request->headers->get('Authorization', ''), 7);
+        if (empty($jwt)) {
+            return null;
+        }
+
+        return array(
+            'jwt' => $jwt,
+        );
+    }
+
+    /**
+     * Returns a user based on the information inside the JSON Web Token depending on the implementation
+     * of the configured user provider.
+     *
+     * When the user provider does not implement the JWTUserProviderInterface it will attempt to load
+     * the user by username with the 'sub' (subject) claim of the JSON Web Token.
+     *
+     * @param array                 $credentials
+     * @param UserProviderInterface $userProvider
+     *
+     * @return UserInterface|null
+     */
+    public function getUser($credentials, UserProviderInterface $userProvider)
+    {
+        try {
+            $jwt = $this->authZeroService->decodeJWT($credentials['jwt']);
+        } catch (CoreException $exception) {
+            // Skip JWT verification exceptions here.
+            // Verification will be done in checkCredentials().
+            return new User('unknown', null, array());
+        }
+
+        if ($userProvider instanceof JWTUserProviderInterface) {
+            return $userProvider->loadUserByJWT($jwt);
+        }
+
+        return $userProvider->loadUserByUsername($jwt->sub);
```

Comment:
Just to be sure ... will `UserProviderInterface` always have a `loadUserByUsername()` method?

Reply:
Yes, it's part of the interface which should be implemented for every UserProvider implementation in Symfony.